File: src/2352_Equal_Row_and_Column_Pairs.py
# 2352_Equal_Row_and_Column_Pairs
# https://leetcode.cn/problems/equal-row-and-column-pairs/submissions/

# 暴力
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        def is_equal(row, col):
            for i in range(len(row)):
                if row[i] != col[i]:
                    return False
            return True

        n = len(grid)
        count = 0
        for r in range(n):
            for c in range(n):
                if is_equal(grid[r], [grid[i][c] for i in range(n)]):
                    count += 1
        return count
    
    
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        n = len(grid)
        cnt = Counter(tuple(r) for r in grid)
        ans = 0
        for col in zip(*grid):
            ans += cnt[col]
        return ans

File: src/2352_Equal_Row_and_Column_Pairs.rs
// 2352_Equal_Row_and_Column_Pairs
// https://leetcode.cn/problems/equal-row-and-column-pairs/
use std::collections::HashMap;
impl Solution {
    pub fn equal_pairs(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let mut cnt = HashMap::new();
    
        for r in &grid {
            *cnt.entry(r.clone()).or_insert(0) += 1;
        }
    
        let mut ans = 0;
        for c in 0..n {
            let col: Vec<_> = grid.iter().map(|r| r[c]).collect();
            ans += cnt.get(&col).unwrap_or(&0);
        }
    
        ans
    }
}


impl Solution {
    pub fn equal_pairs(grid: Vec<Vec<i32>>) -> i32 {
        let mut map = std::collections::HashMap::new();
        for row in grid.iter() {
            *map.entry(row).or_insert(0) += 1;
        }
        let (mut ans, mut temp) = (0, vec![0; grid.len()]);
        for col in 0..grid.len() {
            for row in 0..grid.len() {
                temp[row] = grid[row][col];
            }
            ans += map.get(&temp).unwrap_or(&0);
        }
        ans
    }
}